src/daemon/rpmostree-sysroot-upgrader.c

```diff
@@ -1226,6 +1230,74 @@ rpmostree_sysroot_upgrader_set_kargs (RpmOstreeSysrootUpgrader *self,
   self->kargs_strv = g_strdupv (kernel_args);
 }
 
+static gboolean
+write_history (RpmOstreeSysrootUpgrader *self,
+               OstreeDeployment         *new_deployment,
+               const char               *initiating_command_line,
+               GCancellable             *cancellable,
+               GError                  **error)
+{
+  g_autoptr(GVariant) deployment_variant =
+    rpmostreed_deployment_generate_variant (self->sysroot, new_deployment, NULL,
+                                            self->repo, FALSE, error);
+  if (!deployment_variant)
+    return FALSE;
+
+  g_autofree char *deployment_dirpath =
+    ostree_sysroot_get_deployment_dirpath (self->sysroot, new_deployment);
+  struct stat stbuf;
+  if (!glnx_fstatat (ostree_sysroot_get_fd (self->sysroot),
+                     deployment_dirpath, &stbuf, 0, error))
+    return FALSE;
+
+  g_autofree char *fn =
+    g_strdup_printf ("%s/%ld", RPMOSTREE_HISTORY_DIR, stbuf.st_ctime);
+  if (!glnx_shutil_mkdir_p_at (AT_FDCWD, RPMOSTREE_HISTORY_DIR,
+                               0775, cancellable, error))
+    return FALSE;
+
+  /* Write out GVariant to a file. One obvious question here is: why not keep this in the
+   * journal itself since it supports binary data? We *could* do this, and it would simplify
+   * querying and pruning, but IMO I find binary data in journal messages not appealing and
+   * it breaks the expectation that journal messages should be somewhat easily
+   * introspectable. We could also serialize it to JSON first, though we wouldn't be able to
+   * re-use the printing code in `status.c` as is. Note also the GVariant can be large (e.g.
+   * we include the full `rpmostree.rpmdb.pkglist` in there). */
```

Comment:
What elements of this deployment variant do we use?  The other option is to log those as structured fields right?

Reply:
> What elements of this deployment variant do we use?

Basically almost everything. :) We pass it to the same function in `status.c` that normally prints it. I also want to make use of the pkglist in the future so that we can show the diff from one version to the next.

>  The other option is to log those as structured fields right?

Yeah, I had briefly looked at this. One obvious annoyance is that the field naming convention means we have to do some translation back and forth for the top-level keys. And the values themselves I guess would have to be run through `g_variant_print` and `g_variant_parse`. Still, I go back to whether we should shove that much data in the journal in the first place. (Even though the actual limit is a crazy amount like 768M...). It somewhat reminds me of the D-Bus/`--json` discussion around pkglists and how we ended up blacklisting it, even though it's *mostly* for machines.